### PR TITLE
Create sysfs interface for PFI notification

### PR DESCRIPTION
**Pull request by chadbarth**
Once our main process started using the `/sys/class/gpio/gpio47/value` interface, the gpio helper driver was claiming the PFI GPIO for it's own use. This would cause the pfi.ko to fail to load.

This creates a new `/sys/class/pfi/pfi0/value` interface that mimics the behavior of the old gpio interface. Our app will need to be updated to match this change (but it's OK to merge since we haven't fully integrated the kernel module yet).

### CODE REVIEW

File: pfi.c [hunk 1]
```diff
@@ -26,65 +30,132 @@ static char iface[IFNAMSIZ] = "eth0";
 module_param_string(interface, iface, IFNAMSIZ, 0644);
 MODULE_PARM_DESC(interface, "Interface to disable when interrupt was received.");
 
+static struct class *pfi_class;
+static struct device *pfi_device = NULL;
```
Comment (JNiewiadomski) [inline pfi.c:34]: Any reason why `R34` initializes `pfi_device` but `R33` doesn't initialize `pfi_class`?

File: pfi.c [hunk 2]
```diff
@@ -26,65 +30,132 @@ static char iface[IFNAMSIZ] = "eth0";
...
  * @param  regs   Snapshot of the processor's context before entering interrupt code
  */
-static irq_handler_t pfi_irq_handler(unsigned int irq, void *dev_id, struct pt_regs *regs) {
+static irq_handler_t pfi_irq_handler(unsigned int irq, void *arg, struct pt_regs *regs)
+{
     struct net_device* device;
+    struct kernfs_node *value_sd = arg;
 
     printk(KERN_INFO DRV_NAME ": interrupt received\n");
     device = dev_get_by_name(&init_net, iface);
```
Comment (JNiewiadomski) [inline pfi.c:48]: ~~Nit: Any reason not to delete `R44` and change `R48` to be `net_device * device = dev_get_by_name(&init_net, iface);`?~~

This is C code, isn't it. Ignore my comment.

File: pfi.c [hunk 3]
```diff
@@ -26,65 +30,132 @@ static char iface[IFNAMSIZ] = "eth0";
...
     return (irq_handler_t) IRQ_HANDLED;
 }
 
+static ssize_t value_show(struct device *dev,
+        struct device_attribute *attr, char *buf)
+{
+    ssize_t status;
+    mutex_lock(&pfi_mutex);
+
+    status = sprintf(buf, "%d\n", gpio_get_value(pin));
```
Comment (JNiewiadomski) [inline pfi.c]: Any possibility of having the buffer length passed into `value_show` so that `snprintf` could be used?

Comment (chadbarth) [inline pfi.c]: No, the signature is defined by Linux, but _technically_ we **could** use snprintf. From documentation:

```
- The buffer will always be PAGE_SIZE bytes in length. On i386, this
  is 4096. 
```

Though, we certainly know that `%d` will never expand to anything close to that size.

File: pfi.c (hunk 1 above)
Comment (chadbarth) [inline pfi.c:34]: Yeah, there's a reason. I copied them from different sources. ;-)

Actually, I probably didn't.

I think at one point the code I had possibly called `device_destroy` without necessarily successfully calling `device_create_file` and `device_destroy` properly ignores a NULL device. I don't think that's any longer a possibility, but these should be consistent.

Comment (chadbarth) [inline pfi.c:34]: See https://github.com/Vorne/pfi/pull/2/commits/5f1063bf534b8bf99d92ee406f60c730734ec298

**Comment by JNiewiadomski** [comment]
:+1:

**Pull request merged**